### PR TITLE
Implement cockroachdb autoincrement columns proposal

### PR DESCRIPTION
**Pull request by tomhoule**
This PR implements the proposal on cockroachdb autoincrementing columns for
the introspection and migration engines, and adapts tests where
necessary in the query engine test suite.

This PR is a best effort implementation given the time constraints — it
makes a number of design and implementation compromises to ship the
changes.

**Overview of the changes**

On the cockroachdb connector:

- `@default(autoincrement())` now maps to the default mechanism in
cockroachdb, `DEFAULT unique_rowid()`. Since that makes the column
a 64 bit integer column automatically, this
`@default(autoincrement())` is only supported on `BigInt` fields.`
- A new `@default(sequence())` maps to the usage of sequences in the
database, which is not the recommended default on cockroachdb but
closely matches how postgres autoincrement columns work. The
`sequence()` function can take parameters defined in the PSL proposal.

**Tests**

The main difficulty is with the tests. Now that we map
`@default(autoincrement())` to `DEFAULT unique_rowid()` on cockroachdb
(implicitly through the use of SERIAL), autoincrementing integers in
cockroachdb prisma schemas cannot be `Int` anymore, only `BigInt`. This
makes most of the introspection and migration tests that are portable
between databases not portable to cockroachdb. In this PR, we pick a
subset we still want to test and adapt them to cockroachdb, and ignore
the rest. This is obviously less than ideal.

In the query engine tests, we adapt this similarly, with the additional
difficulty that values produced by unique_rowid() are not stable and do
not start at one. The tests that were not passing are special-cased for
cockroachdb and spread between schemas with `@default(sequence())`
(closely matches postgres @default(autoincrement())) and tests that use
the cockroach default `@default(unique_rowid())`.

An introspection and migrations test setup rewrite initiative is in the
plans.

**Limitations**

- In migrations, we do not diff the settings of existing sequences yet.

closes https://github.com/prisma/prisma/issues/12239
closes https://github.com/prisma/prisma/issues/12113
closes https://github.com/prisma/prisma/issues/12244

### CODE REVIEW

File: libs/datamodel/connectors/sql-datamodel-connector/src/cockroach_datamodel_connector/validations.rs [hunk 1]
```diff
@@ -44,3 +47,18 @@ pub(super) fn inverted_index_validations(index: IndexWalker<'_>, errors: &mut Di
         }
     }
 }
+
+pub(super) fn autoincrement_validations(model: ModelWalker<'_>, errors: &mut Diagnostics) {
```
[the remark this answers is not in the document]
Comment (tomhoule) [reply libs/datamodel/connectors/sql-datamodel-connector/src/cockroach_datamodel_connector/validations.rs:51]: we do, see message above

**Comment by tomhoule** [comment]
I added a test confirming 1244 is fixed.

**Review by Weakky — APPROVED** [review]
All good 👍

File: migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs [hunk 2]
```diff
@@ -25,14 +25,13 @@ use user_facing_errors::{
     UserFacingError,
 };
 
-const ADVISORY_LOCK_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+const ADVISORY_LOCK_TIMEOUT: time::Duration = time::Duration::from_secs(10);
```
Comment (king-of-poppk) [inline migration-engine/connectors/sql-migration-connector/src/flavour/postgres.rs:28]: Is there any chance to make this a user input in the future? I am thinking about `npx prisma migrate deploy --advisory-lock-timeout ...`.

**Pull request merged**